src/simulation.jl

```diff
@@ -143,45 +165,85 @@ end
 is_literal(d::SummationDecapode, var_name::Symbol) = is_literal(d, first(incident(d, var_name, :name)))
 
 is_infer(d::SummationDecapode, var_id::Int) = (d[var_id, :type] == :infer)
 is_infer(d::SummationDecapode, var_name::Symbol) = is_infer(d, first(incident(d, var_name, :name)))
 
-add_stub(stub_name::Symbol, var_name::Symbol) = return Symbol("$(stub_name)_$(var_name)")
+struct InvalidStubException <: Exception
+  stub::Symbol
+end
+
+Base.showerror(io::IO, e::InvalidStubException) = print(io, "Stub \"$(e.stub)\" is invalid")
+
+function add_stub(stub_name::Symbol, var_name::Symbol)
+  # No empty stubs
+  if stub_name == Symbol() || stub_name == NO_STUB_RETURN || !isascii(String(stub_name))
+    throw(InvalidStubException(stub_name))
+  end
+
+  return Symbol("$(stub_name)_$(var_name)")
+end
 
+# ! Warning: Bad behavior if not stub was added and  variable has an "_" itself
 function get_stub(var_name::Symbol)
   var_str = String(var_name)
   idx = findfirst("_", var_str)
-  if(isnothing(idx) || first(idx) == 1)
-    return nothing
+
+  if(isnothing(idx))
+    return NO_STUB_RETURN
+  end
+
+  if(first(idx) == 1)
+    throw(InvalidStubException(var_name))
   end
+
   return Symbol(var_str[begin:first(idx) - 1])
 end
 
-add_inplace_stub(var_name::Symbol) = add_stub(gensim_in_place_stub, var_name)
+add_inplace_stub(var_name::Symbol) = add_stub(GENSIM_INPLACE_STUB, var_name)
+
+const ARITHMETIC_OPS = Set([:+, :*, :-, :/, :.+, :.*, :.-, :./, :^, :.^, :.>, :.<, :.≤, :.≥])
+
+struct InvalidCodeTargetException <: Exception
+  code_target::GenerationTarget
+end
+
+Base.showerror(io::IO, e::InvalidCodeTargetException) = print(io, "Provided code target $(e.code_target) is not yet supported in simulations")
 
-# This will be the function and matrix generation
-function compile_env(d::AbstractNamedDecapode, dec_matrices::Vector{Symbol}, con_dec_operators::Set{Symbol}, code_target::GenerationTarget = CPUTarget())
-  assumed_ops = Set([:+, :*, :-, :/, :.+, :.*, :.-, :./, :^, :.^, :.>, :.<, :.≤, :.≥])
-  defined_ops = Set()
+"""
+    compile_env(d::SummationDecapode, dec_matrices::Vector{Symbol}, con_dec_operators::Set{Symbol}, code_target::GenerationTarget = CPUTarget())
+
+This creates the symbol to function linking for the simulation output. Those run through the `default_dec` backend
+expect both an in-place and an out-of-place variant in that order. User defined operations only support out-of-place.
+"""
+function compile_env(d::SummationDecapode, dec_matrices::Vector{Symbol}, con_dec_operators::Set{Symbol}, code_target::GenerationTarget)
+  defined_ops = deepcopy(con_dec_operators)
 
   defs = quote end
 
   for op in dec_matrices
-    if(op in defined_ops )
+    if op in defined_ops
       continue
     end
 
     quote_op = QuoteNode(op)
-    mat_op = add_stub(gensim_in_place_stub, op)
+    mat_op = add_stub(GENSIM_INPLACE_STUB, op)
 
     # TODO: Add support for user-defined code targets
     default_generation = @match code_target begin
```

Comment:
A later refactor should dispatch on `code_target` directly, instead of checking the type of `code_target` inside the  function itself.

Reply:
Well in this case, I'll only be matching `Backend`s, since these will determine our DEC implementations. Users won't be able to override the entire backend. I like this since this is more concise then multiple functions with multiple dispatch.